XMonad/Hooks/DynamicLog.hs

```diff
@@ -746,6 +764,196 @@ xmobarStripTags tags = strip [] where
 filterOutWsPP :: [WorkspaceId] -> PP -> PP
 filterOutWsPP ws pp = pp { ppSort = (. filterOutWs ws) <$> ppSort pp }
 
+
+-- $further customization
+-- Internally, the status bars are managed through the 'StatusBarConfig' datatype that provides a
+-- convenient abstraction over what a status bar is and how to manage it; and the 'makeStatusBar'
+-- which binds it to an 'XConfig'.
+-- Instead of directly modifying the config, the functions 'statusBarProp' and 'statusBar', create an
+-- appropriate 'StatusBarConfig' and call 'makeStatusBar' on it. This enables the user to have more
+-- flexibility and control over the status bars, without having do the plumbing themselves.
+--
+-- For example, if you want to use property logging as provided by 'statusBarProp', but you want to
+-- customize the action executed to toggle the struts, you could do the following:
+--
+-- > baseSB = statusBarPropConfig "xmobar" myPP -- default key: mod-b
+-- > mySB = withAction (sendMessage ToggleGaps *> sendMessage ToggleStruts) <$> baseSB
+-- > main = xmonad =<< makeStatusBar myConf =<< mySB
+--
+-- The different statusBar functions rely on the following status bar configs:
+--
+-- * 'statusBarProp' relies on 'statusBarPropConfig'
+--
+-- * 'statusBarPropTo' relies on 'statusBarPropConfig'
+--
+-- * 'statusBar' relies on 'statusBarHandleConfig'
+--
+-- * 'statusBar'' relies on 'statusBarHandleConfig''
+--
+-- The main difference is that the functions returning a status bar config do not require a keybinding
+-- and use @mod-b@. Like the toggle action, this can be customized with 'withKey'.
+
+-- | This datataype abstracts a status bar to provide a common interface functions like 'statusBar'
+-- or 'statusBarProp'. Once defined, a status bar can be incorporated in 'XConfig' by using 'makeStatusBar',
+-- which takes care of all the necessary plumbing.
+data StatusBarConfig = StatusBarConfig { sbLogHook     :: X()
```

Comment:
```suggestion
data StatusBarConfig = StatusBarConfig { sbLogHook     :: X ()
```

`X()` feels weird. `()` is a type on its own that is a parameter to the `X` type constructor.